### ISSUE TITLE
Rethink skipUntil and takeUntil

### ISSUE THREAD

**Issue by mattpodwysocki** [ReactiveX/IxJS#42]
**IxJS version:** 2.x

**Code to reproduce:**

**Expected behavior:**

**Actual behavior:**

**Additional information:**

Current implementations of `takeUntil` and `skipUntil` accept `Promise` values as parameters which is problematic given their eager nature.  Instead, we should look into the following overloads which support `AsyncIterator` as well as `() => Promise<any>`.

```typescript
export function skipUntil<T>(
  source: AsyncIterable<T>, 
  other: () => Promise<any>);

export function skipUntil<T>(
  source: AsyncIterable<T>,
  other: AsyncIterable<any>);

export function takeUntil<T>(
  source: AsyncIterable<T>, 
  other: () => Promise<any>);

export function takeUntil<T>(
  source: AsyncIterable<T>,
  other: AsyncIterable<any>);
```

**Comment by OliverJAsh** [comment]
I would also suggest adding a `takeUntil` overload which receives a predicate function, like in RxJava

https://github.com/ReactiveX/IxJS/issues/7#issuecomment-322043165
http://reactivex.io/RxJava/javadoc/rx/Observable.html#takeUntil(rx.functions.Func1)

**Comment by mattpodwysocki** [comment]
Closed as per #75
